nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/TagS3Object.java

```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final long startNanos = System.nanoTime();
+
+        final String bucket = context.getProperty(BUCKET).evaluateAttributeExpressions(flowFile).getValue();
+        final String key = context.getProperty(KEY).evaluateAttributeExpressions(flowFile).getValue();
+        final String newTagKey = context.getProperty(TAG_KEY).evaluateAttributeExpressions(flowFile).getValue();
+        final String newTagVal = context.getProperty(TAG_VALUE).evaluateAttributeExpressions(flowFile).getValue();
+        final String version = context.getProperty(VERSION_ID).evaluateAttributeExpressions(flowFile).getValue();
+
+        final AmazonS3 s3 = getClient();
+
+        SetObjectTaggingRequest r;
+        List<Tag> tags = new ArrayList<>();
+
+        try {
+            if(context.getProperty(APPEND_TAG).asBoolean()) {
+                final GetObjectTaggingRequest gr = new GetObjectTaggingRequest(bucket, key);
+                GetObjectTaggingResult res = s3.getObjectTagging(gr);
+
+                // preserve tags on S3 object, but filter out existing tag keys that match the one we're setting
+                tags = res.getTagSet().stream().filter(t -> !t.getKey().equals(newTagKey)).collect(Collectors.toList());
+            }
+
+            tags.add(new Tag(newTagKey, newTagVal));
+
+            if(StringUtils.isBlank(version)){
+                r = new SetObjectTaggingRequest(bucket, key, new ObjectTagging(tags));
+            } else{
+                r = new SetObjectTaggingRequest(bucket, key, version, new ObjectTagging(tags));
+            }
+            s3.setObjectTagging(r);
+        } catch (final AmazonServiceException ase) {
+            getLogger().error("Failed to tag S3 Object for {}; routing to failure", new Object[]{flowFile, ase});
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        flowFile = setTagAttributes(context, session, flowFile);
+
+        session.transfer(flowFile, REL_SUCCESS);
+        final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+        getLogger().info("Successfully tagged S3 Object for {} in {} millis; routing to success", new Object[]{flowFile, transferMillis});
+    }
+
+    private FlowFile setTagAttributes(ProcessContext context, ProcessSession session, FlowFile flowFile) {
+        final String newTagKey = context.getProperty(TAG_KEY).evaluateAttributeExpressions(flowFile).getValue();
+        final String newTagVal = context.getProperty(TAG_VALUE).evaluateAttributeExpressions(flowFile).getValue();
+
+        if(!context.getProperty(APPEND_TAG).asBoolean()) {
+            flowFile = session.removeAllAttributes(flowFile, Pattern.compile("^s3\\.tag\\..*"));
```

Comment:
I think we should remove existing FlowFile S3 attributes earlier regardless of the `APPEND_TAG` value. Then populate FlowFile attributes using the latest `tags` list at `onTrigger` method.

Current code assumes that the incoming FlowFile has the latest S3 tag attributes. But in case tags are updated elsewhere, the ones retrieved by `GetObjectTaggingRequest` is more reliable.

Also, current code calculates tagKey and tagValue twice. This can be avoided by populating FlowFile attributes from `tags` list at `onTrigger` method.

Reply:
I went back and forth on whether to use the tags pulled from S3 in this processor, or to depend on the conventions in ListS3 processor. Since PutS3Object processor already has the ability to tag objects it is writing to S3, it seems like this TagS3Object processor will be used in conjunction with ListS3 and FetchS3Object (which doesn't actually set tag attributes on the FlowFile).

My thinking is that the existing FlowFile S3 attributes might be the basis for an expression language statement for the new tag. As such, I didn't want to remove those attributes until we've set the new tags in S3. However, it does make sense to reset them to the `tags` list in line 139. Will do.